File: apps/reading/src/api/workstationRecursiveRecordPack.ts
/**
 * Workstation recursive record pack (pure).
 *
 * Operator vision: live in the research workstation; record valuable data,
 * insights, and questions recursively so they inform all prompts.
 *
 * This pure layer packs caller-supplied records for prompt context injection
 * intent. record_persisted and prompts_injected are always false.
 */

export type RecordKind =
  | "insight"
  | "question"
  | "highlight"
  | "finding"
  | "open_thread";

export interface WorkstationRecordItem {
  record_id: string;
  kind: RecordKind;
  text: string;
  /** Optional source asset id. */
  asset_id?: string;
  /** Optional weight 0..1 for prompt prioritization (caller-supplied). */
  weight?: number | null;
}

export interface WorkstationRecursiveRecordPack {
  session_id: string;
  item_count: number;
  by_kind: Record<RecordKind, number>;
  /** Ordered texts ready for prompt context (caller-supplied only). */
  prompt_context_lines: string[];
  /** True when ≥1 item present. */
  pack_ready: boolean;
  /** Always false — pure pack does not write durable records. */
  record_persisted: false;
  /** Always false — pure pack does not inject into live prompts. */
  prompts_injected: false;
  notes: string[];
  authority: "workstation_recursive_record_pack_advisory";
}

const VALID_KINDS = new Set<RecordKind>([
  "insight",
  "question",
  "highlight",
  "finding",
  "open_thread",
]);

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Pack workstation records for recursive prompt-context intent.
 * Never invents content; never persists; never injects prompts.
 */
export function composeWorkstationRecursiveRecordPack(input: {
  session_id: string;
  items: WorkstationRecordItem[];
  /** Optional max lines for prompt_context_lines (positive int when set). */
  max_context_lines?: number | null;
}): WorkstationRecursiveRecordPack {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  if (!Array.isArray(input.items)) {
    throw new Error("items must be an array");
  }

  let maxLines: number | null = null;
  if (input.max_context_lines !== undefined && input.max_context_lines !== null) {
    if (
      typeof input.max_context_lines !== "number" ||
      !Number.isInteger(input.max_context_lines) ||
      input.max_context_lines <= 0
    ) {
      throw new Error("max_context_lines must be a positive integer when set");
    }
    maxLines = input.max_context_lines;
  }

  const notes: string[] = [
    "record_persisted=false — pack intent only",
    "prompts_injected=false — does not mutate live prompts",
    "record texts are caller-supplied only (no invent)",
  ];

  const by_kind: Record<RecordKind, number> = {
    insight: 0,
    question: 0,
    highlight: 0,
    finding: 0,
    open_thread: 0,
  };

  const seenIds = new Set<string>();
  type Scored = { line: string; weight: number; order: number };
  const scored: Scored[] = [];

  for (let i = 0; i < input.items.length; i++) {
    const it = input.items[i];
    if (!it || typeof it !== "object") {
      throw new Error(`items[${i}] must be an object`);
    }
    const record_id = requireNonEmpty(it.record_id, `items[${i}].record_id`);
    if (seenIds.has(record_id)) {
      throw new Error(`duplicate record_id: ${record_id}`);
    }
    seenIds.add(record_id);

    const kind = it.kind;
    if (!VALID_KINDS.has(kind as RecordKind)) {
      throw new Error(
        `items[${i}].kind must be insight|question|highlight|finding|open_thread`,
      );
    }
    const k = kind as RecordKind;
    by_kind[k] += 1;

    const text = requireNonEmpty(it.text, `items[${i}].text`);
    let weight = 0.5;
    if (it.weight !== undefined && it.weight !== null) {
      if (
        typeof it.weight !== "number" ||
        !Number.isFinite(it.weight) ||
        it.weight < 0 ||
        it.weight > 1
      ) {
        throw new Error(`items[${i}].weight must be finite in [0, 1] when set`);
      }
      weight = it.weight;
    }

    let assetPart = "";
    if (it.asset_id != null) {
      const aid = requireNonEmpty(it.asset_id, `items[${i}].asset_id`);
      assetPart = ` @${aid}`;
    }

    scored.push({
      line: `[${k}]${assetPart} ${text}`,
      weight,
      order: i,
    });
  }

  // Higher weight first; stable by input order for ties.
  scored.sort((a, b) => {
    if (b.weight !== a.weight) return b.weight - a.weight;
    return a.order - b.order;
  });

  let prompt_context_lines = scored.map((s) => s.line);
  if (maxLines !== null && prompt_context_lines.length > maxLines) {
    notes.push(
      `max_context_lines=${maxLines} — truncated from ${prompt_context_lines.length}`,
    );
    prompt_context_lines = prompt_context_lines.slice(0, maxLines);
  }

  const item_count = input.items.length;
  const pack_ready = item_count >= 1;
  if (!pack_ready) {
    notes.push("pack_ready=false — empty items (no invent records)");
  } else {
    notes.push(
      `pack_ready=true · items=${item_count} · context_lines=${prompt_context_lines.length}`,
    );
  }
  notes.push("record_persisted=false");
  notes.push("prompts_injected=false");

  return {
    session_id,
    item_count,
    by_kind,
    prompt_context_lines,
    pack_ready,
    record_persisted: false,
    prompts_injected: false,
    notes,
    authority: "workstation_recursive_record_pack_advisory",
  };
}

export function formatWorkstationRecursiveRecordPackSummary(
  p: WorkstationRecursiveRecordPack,
): string {
  return (
    `record pack ${p.session_id} · items=${p.item_count} · ` +
    `ready=${p.pack_ready} · record_persisted=false · prompts_injected=false`
  );
}

File: substrate/workstation_recursive_record_pack.py
"""Workstation recursive record pack (pure, advisory).

Packs caller-supplied insights/questions/highlights for prompt context.
record_persisted and prompts_injected are always False.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any, Literal

RecordKind = Literal[
    "insight", "question", "highlight", "finding", "open_thread"
]

VALID_KINDS = frozenset(
    {"insight", "question", "highlight", "finding", "open_thread"}
)


class WorkstationRecursiveRecordPackError(ValueError):
    """Fail-closed validation for workstation record pack."""


@dataclass(frozen=True)
class WorkstationRecursiveRecordPack:
    session_id: str
    item_count: int
    by_kind: dict[str, int]
    prompt_context_lines: tuple[str, ...]
    pack_ready: bool
    record_persisted: bool
    prompts_injected: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "item_count": self.item_count,
            "by_kind": dict(self.by_kind),
            "prompt_context_lines": list(self.prompt_context_lines),
            "pack_ready": self.pack_ready,
            "record_persisted": False,
            "prompts_injected": False,
            "notes": list(self.notes),
            "authority": "workstation_recursive_record_pack_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise WorkstationRecursiveRecordPackError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_workstation_recursive_record_pack(
    *,
    session_id: object,
    items: object,
    max_context_lines: object | None = None,
) -> WorkstationRecursiveRecordPack:
    """Pack workstation records. Never persists or injects prompts."""
    sid = _require_nonempty(session_id, field="session_id")
    if not isinstance(items, list):
        raise WorkstationRecursiveRecordPackError("items must be an array")

    max_lines: int | None = None
    if max_context_lines is not None:
        if (
            isinstance(max_context_lines, bool)
            or not isinstance(max_context_lines, int)
            or max_context_lines <= 0
        ):
            raise WorkstationRecursiveRecordPackError(
                "max_context_lines must be a positive integer when set"
            )
        max_lines = max_context_lines

    notes: list[str] = [
        "record_persisted=false — pack intent only",
        "prompts_injected=false — does not mutate live prompts",
        "record texts are caller-supplied only (no invent)",
    ]
    by_kind: dict[str, int] = {
        "insight": 0,
        "question": 0,
        "highlight": 0,
        "finding": 0,
        "open_thread": 0,
    }
    seen: set[str] = set()
    scored: list[tuple[float, int, str]] = []

    for i, it in enumerate(items):
        if not isinstance(it, dict):
            raise WorkstationRecursiveRecordPackError(
                f"items[{i}] must be an object"
            )
        rid = _require_nonempty(it.get("record_id"), field=f"items[{i}].record_id")
        if rid in seen:
            raise WorkstationRecursiveRecordPackError(
                f"duplicate record_id: {rid}"
            )
        seen.add(rid)
        kind = it.get("kind")
        if kind not in VALID_KINDS:
            raise WorkstationRecursiveRecordPackError(
                f"items[{i}].kind must be insight|question|highlight|finding|open_thread"
            )
        by_kind[kind] += 1  # type: ignore[index]
        text = _require_nonempty(it.get("text"), field=f"items[{i}].text")

        weight = 0.5
        w = it.get("weight")
        if w is not None:
            if isinstance(w, bool) or not isinstance(w, (int, float)):
                raise WorkstationRecursiveRecordPackError(
                    f"items[{i}].weight must be finite in [0, 1] when set"
                )
            wf = float(w)
            if not math.isfinite(wf) or wf < 0 or wf > 1:
                raise WorkstationRecursiveRecordPackError(
                    f"items[{i}].weight must be finite in [0, 1] when set"
                )
            weight = wf

        asset_part = ""
        if it.get("asset_id") is not None:
            aid = _require_nonempty(
                it.get("asset_id"), field=f"items[{i}].asset_id"
            )
            asset_part = f" @{aid}"

        line = f"[{kind}]{asset_part} {text}"
        scored.append((weight, i, line))

    scored.sort(key=lambda t: (-t[0], t[1]))
    lines = [t[2] for t in scored]
    if max_lines is not None and len(lines) > max_lines:
        notes.append(
            f"max_context_lines={max_lines} — truncated from {len(lines)}"
        )
        lines = lines[:max_lines]

    item_count = len(items)
    pack_ready = item_count >= 1
    if not pack_ready:
        notes.append("pack_ready=false — empty items (no invent records)")
    else:
        notes.append(
            f"pack_ready=true · items={item_count} · context_lines={len(lines)}"
        )
    notes.append("record_persisted=false")
    notes.append("prompts_injected=false")

    return WorkstationRecursiveRecordPack(
        session_id=sid,
        item_count=item_count,
        by_kind=by_kind,
        prompt_context_lines=tuple(lines),
        pack_ready=pack_ready,
        record_persisted=False,
        prompts_injected=False,
        notes=tuple(notes),
        authority="workstation_recursive_record_pack_advisory",
    )


__all__ = [
    "WorkstationRecursiveRecordPack",
    "WorkstationRecursiveRecordPackError",
    "compose_workstation_recursive_record_pack",
]
